agg-ast/schema_derivation/src/match_schema_derivation.rs

```diff
@@ -1135,30 +1285,181 @@ impl MatchConstrainSchema for Expression {
+            if let Some(options) = &r.options {
+                // options can be NULLISH no matter the current null_behavior
+                handle_arg!(*options.as_ref(), STRING_OR_NULLISH.clone());
+            }
+            Ok(())
+        }
+
+        fn match_derive_to_string(u: &UntaggedOperator, state: &mut ResultSetState) -> Result<()> {
+            if u.args.is_empty() {
+                return Err(Error::NotEnoughArguments(u.op.to_string()));
+            }
+            if let Expression::Ref(reference) = u.args[0].clone() {
+                let schema = match state.null_behavior {
+                    Satisfaction::Not => Schema::AnyOf(set! {
+                        Schema::Atomic(Atomic::BinData),
+                        Schema::Atomic(Atomic::Boolean),
+                        Schema::Atomic(Atomic::Double),
+                        Schema::Atomic(Atomic::Decimal),
+                        Schema::Atomic(Atomic::Integer),
+                        Schema::Atomic(Atomic::Long),
+                        Schema::Atomic(Atomic::ObjectId),
+                        Schema::Atomic(Atomic::String),
+                        Schema::Atomic(Atomic::Date),
+                    }),
+                    Satisfaction::May => Schema::AnyOf(set! {
+                        Schema::Atomic(Atomic::BinData),
+                        Schema::Atomic(Atomic::Boolean),
+                        Schema::Atomic(Atomic::Double),
+                        Schema::Atomic(Atomic::Decimal),
+                        Schema::Atomic(Atomic::Integer),
+                        Schema::Atomic(Atomic::Long),
+                        Schema::Atomic(Atomic::ObjectId),
+                        Schema::Atomic(Atomic::String),
+                        Schema::Atomic(Atomic::Date),
+                        Schema::Atomic(Atomic::Null),
+                        Schema::Missing,
+                    }),
+                    Satisfaction::Must => NULLISH.clone(),
+                };
+                intersect_if_exists(&reference, state, schema);
+            } else {
+                u.args[0].match_derive_schema(state)?;
+            }
+            Ok(())
+        }
+
+        fn match_derive_trim(t: &Trim, state: &mut ResultSetState) -> Result<()> {
+            let arg_schema = match state.null_behavior {
+                Satisfaction::Not => Schema::Atomic(Atomic::String),
+                Satisfaction::May | Satisfaction::Must => STRING_OR_NULLISH.clone(),
+            };
+            macro_rules! handle_arg {
+                ($input:expr, $sch:expr) => {
+                    if let Expression::Ref(ref reference) = $input {
+                        intersect_if_exists(reference, state, $sch);
+                    } else {
+                        $input.match_derive_schema(state)?;
+                    }
+                };
+            }
+            handle_arg!(*t.input, arg_schema.clone());
+            if let Some(chars) = &t.chars {
```

Comment:
we've had this discussion before and I think its pretty outside the scope of this epic. But we could technically be more precise in these situations. What if one of the args must not be null (eg it is a literal string). Then, when the null behavior is "must be null", we might be able to infer a field reference _must_ be null. 

Nothing to see here, just noting it. Perhaps something that we should make sure to include in the closeout doc for posterity.

Reply:
I think so. We could be more precise, but a lot of the precision will result in cartesian blowups. But maybe that's fine for operators that have 1-4 args.